mds/hardlink: Create a referrent inode

kotreshhr opened the pull request:
1. Creates a secondary inode and stores the backtrace on link local

Pending:
1. Do the same on remote linking
2. Fix caching

Fixes: https://tracker.ceph.com/issues/54205
Signed-off-by: Kotresh HR <khiremat@redhat.com>





<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

kotreshhr commented:
jenkins test make check

kotreshhr commented:
**PENDING TASKS**
- [ ] Fix multi mds i.e., do the same on remote linking. Currently this is done only for link_local i.e., single mds
- [ ] Fix Caching - Load referent CInode into mdcache.
- [ ] Locks - Take appropriate locks on referent CInode to update backtrace and fields if any.
- [ ] Custom CInode to store backtrace. Currenlty it's full blown CInode structure which contains unused fields.
- [ ] Store reverse links. Point to all the referent inodes from primary link (Store list of referent inodes in primary CInode)

mchangir commented on src/mds/CDentry.cc:
```diff
@@ -299,6 +299,20 @@ void CDentry::push_projected_linkage()
+  //flushing dirty_inode in try_to_expire
+  ref_inode->push_projected_parent(this);
+  //ref_inode->set_remote_ino(remote_ino);
+
```
is this commented code intentional ?

mchangir commented on src/mds/Server.cc:
```diff
@@ -8049,11 +8059,33 @@ void Server::_unlink_local(MDRequestRef& mdr, CDentry *dn, CDentry *straydn)
+      pri.inode->update_backtrace();
+      le->metablob.add_primary_dentry(straydn, ref_in, true, true);
+
+      mdcache->predirty_journal_parents(mdr, &le->metablob, in, dn->get_dir(), PREDIRTY_DIR, -1);
+      mdcache->predirty_journal_parents(mdr, &le->metablob, in, 0, PREDIRTY_PRIMARY);
```
please use `nullptr` for NULL pointer types - in this case for the parent dir pointer

mchangir commented on src/mds/Server.cc:
```diff
@@ -9310,15 +9319,28 @@ void Server::_rename_prepare(MDRequestRef& mdr,
         spi = pi.inode.get();
       }
+      //src_refi
+      /*
+      if (src_refi->is_auth()) {
+	auto pri = src_refi->project_inode(mdr);
+	pri.inode->version = src_refi->pre_dirty();
+	pri.inode->update_backtrace();
+	//TODO Do we need to update ctime of the referent inode ?
+      }*/
```
commented code ?

mchangir commented on src/messages/MMDSCacheRejoin.h:
```diff
@@ -97,6 +97,7 @@ class MMDSCacheRejoin final : public MMDSOp {
       first(f), alternate_name(altn), ino(pi), remote_ino(ri), remote_d_type(rdt), nonce(n), lock(l) {}
     bool is_primary() const { return ino > 0; }
     bool is_remote() const { return remote_ino > 0; }
+    bool is_referent() const { return remote_ino > 0; }
```
just wanted to get clarification for **is_remote** and **is_referent** referring to the same data member i.e. `remote_ino`

mchangir reviewed (commented):
just primary review comments

kotreshhr commented:
jenkins test submodules

petrutlucian94 commented:
jenkins test windows

petrutlucian94 commented:
I think ``LibCephFS.HardlinkNoOriginal`` hanged or crashed on Windows.

FWIW, the Windows job is running ``ceph_test_libcephfs``, unlike the ``make check`` job. This can help uncovering bugs early.

leonid-s-usov commented on src/include/cephfs/types.h:
```diff
@@ -547,6 +553,8 @@ struct inode_t {
 
   bufferlist fscrypt_last_block;
 
+  inodeno_t remote_ino = 0; // referent inode - remote inode link
```
Do I understand correctly that `remote_ino` and `referent_inodes` aren't ever going to be used simultaneously? If so, then these fields should be under a union to not waste space.
Alternatively, the remote_ino could be a single member of the collection if the inode represents a referent. The collection could be called `remote_inodes` and treated respectively to the inode type.

leonid-s-usov reviewed (commented):
I want to clarify why `is_remote` and `is_referrent` are exclusive dentry states. I think that in both cases we should get `true` for `is_remote`. That would save us numerous changes where we had to modify conditions as `is_remote() || is_referrent()`.
On the rare occasions when we do want to work on the referent inode which wasn't previously available for remote dentries, we could check for whether it's not null.

My second question is about why we need a referent inode at all, what is its role beyond storing the remote inode number which our dentry already holds?

kotreshhr commented:
jenkins test make check

vshankar reviewed (commented):
Still going through majority of the changes. Leaving my initial findings for now.

vshankar commented:
Last meeting @leonid-s-usov bought up the point about the CInode structure being used for the referent inode vs having a trimmed down inode structure with just enough details (structure members) that's required (or maybe even having those details in the CDentry structure). I want to emphasise the point that the main reasons to choose CInode was a) Possible requirement of the CInode structure when faster clones (layering) support is designed/implemented and b) Easy to work with many MDS internal APIs that accept CInode. Needless to say, the former was the main driving factor for choosing to use CInode. @kotreshhr??

Note that using CInode for referent inodes isn't cast in stone - we had the option to choose a "referent specific" structure and later move to CInode if layering support drives the need. I don't want to call is premature optimization too, but the point is that we are flexible to use what's required when designing layering support.

(I couldn't actively participate in last week's calls since I was multitasking with folks pinging me about downstream stuff).

leonid-s-usov commented:
> Last meeting @leonid-s-usov bought up the point about the CInode structure being used for the referent inode vs having a trimmed down inode structure with just enough details (structure members) that's required (or maybe even having those details in the CDentry structure). I want to emphasise the point that the main reasons to choose CInode was a) Possible requirement of the CInode structure when faster clones (layering) support is designed/implemented and b) Easy to work with many MDS internal APIs that accept CInode. Needless to say, the former was the main driving factor for choosing to use CInode. @kotreshhr??
> 
> Note that using CInode for referent inodes isn't cast in stone - we had the option to choose a "referent specific" structure and later move to CInode if layering support drives the need. I don't want to call is premature optimization too, but the point is that we are flexible to use what's required when designing layering support.
> 
> (I couldn't actively participate in last week's calls since I was multitasking with folks pinging me about downstream stuff).

I'd like to clarify that there are two different aspects here: the usage of CInode in-memory after loading the new type of dentry, and the on-disk representation of the dentry.

My comment was mainly about the latter: I didn't see any reason to serialize an empty CInode structure, especially given the fact that even its snapshot-related fields aren't being used as of now. However, it doesn't prevent us from instantiating a CInode into the cache when loading such a dentry.

There was a research item for Kotresh to clarify whether the snapshot-related CInode fields will be needed for the referent inode. If that will turn out to be the case, then my second comment applies about potentially serializing just those fields and then again populating a full-size CInode structure with the relevant info when loading of such a dentry.

leonid-s-usov commented:
Another comment I had was about using the `inode` in the cdentry (both in memory and on disk) to store the referent inode number. It doesn't have to be as radical as also returning the referent inode for `get_inode()`. It could be something like

```
    bool is_primary() const { return remote_ino == 0 && inode != nullptr; }
    bool is_remote() const { return remote_ino > 0; }
    bool is_rererent() const { return remote_ino > 0 && inode != nullptr; }
    bool is_null() const { return remote_ino == 0 && inode == nullptr; }
    CInode* get_inode() const { remote_ino == 0 ? inode : nullptr; }
    CInode* get_referent_inode() const { remote_ino == 0 ? nullptr : inode; }
```

with this approach, the existing code paths won't be affected (apart from the fact that here `is_remote` will be true for both old style and new style remote nodes, which is a different topic) but we won't have to waste another 8 bytes per cdentry for the redundant `referent_ino` field.

kotreshhr commented:
> Another comment I had was about using the `inode` in the cdentry (both in memory and on disk) to store the referent inode number. It doesn't have to be as radical as also returning the referent inode for `get_inode()`. It could be something like
> 
> ```
>     bool is_primary() const { return remote_ino == 0 && inode != nullptr; }
>     bool is_remote() const { return remote_ino > 0; }
>     bool is_referent() const { return remote_ino > 0 && inode != nullptr; }
>     bool is_null() const { return remote_ino == 0 && inode == nullptr; }
>     CInode* get_inode() const { remote_ino == 0 ? inode : nullptr; }
>     CInode* get_referent_inode() const { remote_ino == 0 ? nullptr : inode; }
> ```
> 
> with this approach, the existing code paths won't be affected (apart from the fact that here `is_remote` will be true for both old style and new style remote nodes, which is a different topic) but we won't have to waste another 8 bytes per cdentry for the redundant `referent_ino` field.

The `dnl->inode` is dedicated to target inode in the code. It is filled in both the cases where the inode is local or the remote i.e., both for primary link file where the `remote_ino == 0` and the secondary link file where the `remote_ino > 0`. The `dnl->inode` is reserved for target inode's CInode pointer. It can't be overloaded as referent inode.
         
         Please look at the following code:
         https://github.com/ceph/ceph/blob/main/src/mds/Server.cc#L7423
         https://github.com/ceph/ceph/blob/main/src/mds/Server.cc#L7554

leonid-s-usov commented:
> Please look at the following code:
> 
> https://github.com/ceph/ceph/blob/main/src/mds/Server.cc#L7423 https://github.com/ceph/ceph/blob/main/src/mds/Server.cc#L7554

Alright, I see that ultimately you are referring to https://github.com/ceph/ceph/blob/main/src/mds/CDentry.cc#L272C1-L272C1. Indeed then, you will need a separate CInode* for the referent inode. As for the on-disc structure, I see that it's already highly dynamic based on the prior knowledge, and currently, remote dentries only encode a single inode number so most probably this optimization won't be possible

kotreshhr commented:
Rebased resolving the conflict. Also added commits handling the replica during multi-mds link/unlink and reverse link updation on multi-mds link/unlink

vshankar commented:
@kotreshhr - https://github.com/ceph/ceph/compare/main...vshankar:ceph:wip-sample-test-hardlink

see: test_hardlink_remote_auth() - right now the inode from the auth mds of the inode is dumped, just do the same from the replica mds. Its likely that its the mdlog->flush in the auth mds that's fixing up the replicas.

kotreshhr commented:
Removing the stale label. Getting back to work on this.

kotreshhr commented:
jenkins test make check

kotreshhr commented:
https://pulpito.ceph.com/khiremat-2024-11-21_13:43:13-fs-wip-khiremat-49945-referent-inode-multimds-Nov212024-testing-default-smithi/

A lot of failures - Needs to look into them

kotreshhr commented:
jenkins test make check

kotreshhr commented:
jenkins test make check

hkadam134 commented:
This PR is under test in [https://tracker.ceph.com/issues/69611](https://tracker.ceph.com/issues/69611).

hkadam134 commented:
This PR is under test in [https://tracker.ceph.com/issues/69614](https://tracker.ceph.com/issues/69614).

batrick reviewed (changes requested):
superseded by #61945 ?

kotreshhr commented:
> superseded by #61945 ?

Yes, that's correct!

vshankar commented:
> > superseded by #61945 ?
> 
> Yes, that's correct!

Let's close this PR then. The global snap realm changes would be different PR anyway.

kotreshhr commented:
> > > superseded by #61945 ?
> > 
> > 
> > Yes, that's correct!
> 
> Let's close this PR then. The global snap realm changes would be different PR anyway.

Yes, I will close this once I raise the global snaprealm PR. I will convert this to draft for now
